### PR TITLE
Extend variable-length encoding to Uint types

### PR DESCRIPTION
**Pull request by umanwizard**
I originally thought we could leave these out because they're rarely used, but there are two cases I forgot about:

1. introspection sources,
2. `WITH PROGRESS` in sources

### Motivation

  * This PR fixes a recognized bug.

    Materialize uses too much memory for encoding unsigned integers

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
@MaterializeInc/testing folks might want to give this a pass
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Causes Materialize to use less memory in some cases

### CODE REVIEW

File: src/repr/src/row.rs [hunk 1]
```diff
@@ -432,7 +432,7 @@ enum Tag {
     // i64 nanoseconds. (Note that Materialize does not support leap
     // seconds, but this module does).
     CheapTimestampTz,
-    // The remaining tags are for variable-length signed integer encoding.
+    // The next several tags are for variable-length signed integer encoding.
     // The basic idea is that `NonNegativeIntN_K` is used to encode a datum of type
     // IntN whose actual value is positive or zero and fits in K bytes, and similarly for
```
Comment (vmarcos) [inline src/repr/src/row.rs:437]: Do you mean here `K bytes` or `K bits`?

File: src/repr/src/row.rs [hunk 2]
```diff
@@ -482,21 +482,49 @@ enum Tag {
...
+    UInt64_24,
+    UInt64_32,
+    UInt64_40,
+    UInt64_48,
+    UInt64_56,
+    UInt64_64,
 }
 
 impl Tag {
     fn actual_int_length(self) -> Option<usize> {
```
Comment (vmarcos) [inline src/repr/src/row.rs:513]: In there a reason for why we chose to not add an `#[inline]` annotation here?

**Review by vmarcos — COMMENTED** [review]
This looks good to me overall, but it would be ideal if we could extend the unit tests to visit the new code paths. In particular, there is an opportunity to add unsigned integers to `test_datum_sizes` as well as to `miri_test_round_trip`.

In the latter, we should add to the fixed `vec!` and also extend `ScalarType::interesting_datums` to cover all the size tags.

File: src/repr/src/row.rs (hunk 2 above)
Comment (umanwizard) [reply src/repr/src/row.rs:513]: No, I haven't measured it either way.

**Comment by nrainer-materialize** [comment]
I triggered a [nightly build](https://buildkite.com/materialize/nightlies/builds/5233) and will assess the results when they are ready (there will unfortunately be some unrelated failures).

**Review by vmarcos — APPROVED** [review]
Thanks for the additional tests and revisions! Looks good from my side, pending @nrainer-materialize's assessment of the nightly run.

File: src/repr/src/row.rs (hunk 2 above)
Comment (vmarcos) [reply src/repr/src/row.rs:513]: Definitely not a blocker for merging, so please feel free to look into this item afterwards if it is better time-wise.

**Pull request merged**